JitsiMediaDevices.js

```diff
@@ -54,26 +110,58 @@ const JitsiMediaDevices = {
      */
     isDeviceChangeAvailable(deviceType) {
         return RTC.isDeviceChangeAvailable(deviceType);
-    },
+    }
 
     /**
-     * Returns true if user granted permission to media devices.
+     * Checks if the permission for the given device was granted.
+     *
      * @param {'audio'|'video'} [type] - type of devices to check,
      *      undefined stands for both 'audio' and 'video' together
-     * @returns {boolean}
+     * @returns {Promise<boolean>}
      */
     isDevicePermissionGranted(type) {
-        const permissions = RTC.getDeviceAvailability();
-
-        switch (type) {
-        case MediaType.VIDEO:
-            return permissions.video === true;
-        case MediaType.AUDIO:
-            return permissions.audio === true;
-        default:
-            return permissions.video === true && permissions.audio === true;
-        }
-    },
+        return new Promise(resolve => {
+            // Shortcut: first check if we already know the permission was
+            // granted.
+            if (type in this._grantedPermissions) {
+                resolve(this._grantedPermissions[type] === true);
+
+                return;
+            }
+
+            // Check using the Permissions API.
+            this._permissionsApiSupported.then(supported => {
+                if (!supported) {
+                    resolve(false);
+
+                    return;
+                }
+
+                const promises = [];
+
+                switch (type) {
+                case MediaType.VIDEO:
+                    promises.push(
+                        navigator.permissions.query({ name: 'camera' }));
+                    break;
+                case MediaType.AUDIO:
+                    promises.push(
+                        navigator.permissions.query({ name: 'microphone' }));
+                    break;
+                default:
+                    promises.push(
+                        navigator.permissions.query({ name: 'camera' }));
+                    promises.push(
+                        navigator.permissions.query({ name: 'microphone' }));
+                }
+
+                Promise.all(promises).then(
+                    r => resolve(r.every(Boolean)),
+                    () => resolve(false)
```

Comment:
You think it's worthwhile to update the known permissions at this point? Maybe a premature optimization.

Reply:
They will be updated as soon as streams are created, so I think having a single place where that info comes from makes the code less involved. WDYT?

Reply:
Sounds fair and right.